### DISCUSSION TITLE
How to clone two single-node models into multiple nodes without success

### DISCUSSION THREAD

**Discussion by meng25meng** [iiasa/message_ix#818] [MESSAGE modeling]
I want to set up power transmission of two single-node model provinces through one-way transmission, according to the second method in the tutorial, but at the beginning, it is difficult to clone the single-node model, and it is always impossible to read the two models. My nodes do not want to create new ones, which are all from the single-node model. So far cloning has been unsuccessful.

```python
import pandas as pd
from itertools import product
import ixmp
import message_ix

from message_ix.util import make_df

%matplotlib inline
# Loading the modeling platform
mp = ixmp.Platform()
model1 = "Anhui energy model"
scenario1 = "baseline2"
base = message_ix.Scenario(mp, model1, scenario1)
model2 = "Xinjiang energy model"
scenario2= "baselinexinjiang_0.05"
base = message_ix.Scenario(mp, model2, scenario2)
base.has_solution()
new_scenario = "multinode"
scen = base.clone(scenario=new_scenario, keep_solution=False)
scen = base.clone(scenario=new_scenario, keep_solution=False)
# Checking out the scenario for editing
scen.check_out()
scen.set("node")
0       World
1    Xinjiang
dtype: object**
```

**Comment by glatterf42** [comment]
Hi @meng25meng, thanks for your question. In your code snippet, you define a variable `base` as a `message_ix.Scenario` with model "Anhui energy model" and scenario "baseline2", but then you re-define the variable `base` as a `message_ix.Scenario` with model "Xinjiang energy model" and scenario "baselinexinjiang_0.05". If you want to load both `Scenario`s at the same time, you will need to use different variables for that.

However, I'm not sure this is what you want to do. Please correct me if I'm wrong, but to me it sounds like you want to have a single `Scenario` with two nodes in it that should be connected via power/energy trade/transmission. If this interpretation is correct, you don't need to define two `Scenario`s, you just need to add both nodes (or regions) to one `Scenario`. Please see step 1.1 of [this tutorial](https://github.com/iiasa/message_ix/blob/main/tutorial/westeros/westeros_multinode_energy_trade.ipynb) to learn how to do this.

Please note: I think this is a question about how to use message_ix more than it is an issue with the code. So I will convert this to a discussion.

**Reply by meng25meng** [reply]
Yes, I would like to adopt the second transmission mode in the tutorial to establish the transmission line between Anhui and Xinjiang. Their single-node models have been established, and I have referred to the tutorial at the beginning. I think it is necessary to clone the single-node models and put the two provinces in the same scenario. Following the method you mentioned is not feasible, so we need to add nodes in the next step. Re-adding the nodes, then I want to use their own single-node models to store their data, so do I need to clone their single-node models and add data separately?

**Reply by glatterf42** [reply]
I'm sorry, what to do you mean by "following the method is not feasible"?
I think in order to make your idea work, you will have to combine pieces of the `westeros_baseline` and `westeros_multinode_energy_trade` tutorials. You could for example start like this:

```python
import ixmp

import message_ix

%matplotlib inline

mp = ixmp.Platform()
scenario = message_ix.Scenario(
    mp=mp,
    model="Two Nodes with Energy transmission",
    scenario="baseline",
    version="new",
)

# Let's follow the westeros baseline tutorial by defining basic characteristics of the
# model, including time, space, and the energy system structure.

history = [2020]
model_horizon = [2030, 2040, 2050]
scenario.add_horizon(year=history + model_horizon, firstmodelyear=model_horizon[0])

scenario.add_spatial_sets({"country": {"China": {"state": ["Anhui", "Xinjiang"]}}})

# You could define the energy system structure here, for example:
# scenario.add_set("commodity", ["electricity", "light"])
# scenario.add_set("level", ["secondary", "final", "useful"])
# scenario.add_set("technology", ["coal_ppl", "wind_ppl", "grid", "bulb"])
# scenario.add_set("mode", "standard")
```

This will already give you
```python
> scenario.set("node")
0       World
1       Anhui
2    Xinjiang
3       China
dtype: object
```

Next, you would need to define demand for your regions because in the end, the model will optimize such that the demand is being met. For this, you can follow the `westeros_baseline` tutorial, but when you define your demand like this...
```python
light_demand = pd.DataFrame(
    {
        "node": country,
        "commodity": "light",
        "level": "useful",
        "year": model_horizon,
        "time": "year",
        "value": (demand_per_year * gdp_profile).round(),
        "unit": "GWa",
    }
)
```
...remember to define it for both nodes you are interested in by replacing `country` with the name of the node you want to define. Alternatively, you can define the demand (and everything else) for only one node and then copy the data over to the other node using `scenario.rename()` as detailed in part 1 of the `multinode` tutorial. The same goes for the definition of the various other model parameters like `input` and `output` of the technologies, their `technical_lifetime`, `activity` constraints, etc.
Finally, to establish direct trade between `Anhui` and `Xinjiang`, you can follow along with part 3 of the `multinode` tutorial.

**Reply by meng25meng** [reply]
Thank you for your answer. My previous practice is different from what you said. I set up a single node for each province. The reason for this is that I need to set up power transmission between up to 8 provinces, and if I divide it into 8 nodes in the basic tutorial, I may need to use a commercial solver and cannot afford to do this. Therefore, I am referring to the baseline tutorial to establish a single-node model, which is to demonstrate Anhui and Xinjiang, as follows:
```python
model = "Anhui energy model"
scen = "baseline2"
annot = "developing a stylized energy system model for illustration and testing" 

scenario = message_ix.Scenario(mp, model, scen, version='new', annotation=annot)

# Time and Spatial Detail
#The model includes the time periods 2010, 2020, 2030 and 2040.（5年时间片）
history = [2005,2010,2015,2020]
horizon = [2025, 2030, 2035, 2040, 2045, 2050, 2055, 2060]
scenario.add_horizon(
    year= history + horizon,
    firstmodelyear=horizon[0]
)
history

model = "Xinjiang energy model"
scen = "baseline1"
annot = "developing a stylized energy system model for illustration and testing" 

scenario = message_ix.Scenario(mp, model, scen, version='new', annotation=annot)

# Time and Spatial Detail
#The model includes the time periods 2010, 2020, 2030 and 2040.（5年时间片）
history = [2005, 2010, 2015, 2020]
horizon = [2025, 2030, 2035, 2040, 2045, 2050, 2055, 2060]
scenario.add_horizon(
    year= history + horizon,
    firstmodelyear=horizon[0]
)
horizon
```
Then want to build two single-node model of power transmission, refer to the multi-node model of the tutorial, in this multi-node model, there is no need to add new nodes, so tutorial 1.2 here do not know how to represent, respectively clone their own model?

**Reply by glatterf42** [reply]
Sorry for your inconvenience while trying to avoid a commercial solver! We are aware of the costs of offering only the GAMS solver and want to explore alternatives in the near future. At the moment, every call to `Scenario.solve()` will trigger a run of GAMS, but I don't know how much of that can be covered by free licenses. Have you explored using the HiGHS solver instead of CPLEX? This solver is free to use and would probably what we will start experimenting with.

I don't know if you can have transmission between different `Scenario`s, I don't think so, but maybe @khaeru can clarify. As far as I know, a `Scenario` is a self-contained model of the world (or some smaller part of it), so all interaction between regions will happen inside of a `Scenario`. If you create multiple `Scenario`s with one region (or node) each, message_ix will think that you want to model different regions in an isolated way.

**Reply by meng25meng** [reply]
Thank you for your answer, do you mean that this power transmission can represent the internal nodes of each system model and not between two single-node models? I might have a little trouble with my code modeling work on that one. If you switch to HIGHS solver can you build a multi-node model as you mentioned above and then add the transfer constraints to it? If so, is there a tutorial on commutation solvers? Or is this difficult? I'm not a professional programmer, and there are some things I don't understand. Please answer. thank you.

**Reply by glatterf42** [reply]
As far as I know, your interpretation is correct that the power transmission can only include internal nodes of each system model.

So for the GAMS-problem, you have a few options:

1. Switch to the HiGHS solver
    - This possibility has been discussed in https://github.com/iiasa/message_ix/discussions/794 before; you can check that thread to find how you would go about this
    - We currently don't have any kind of test, tutorial, or documentation for this
    - We will want to explore this option ourselves at some point, so any contribution you could make would be highly welcomed 
2. Sign up to the IIASA capacity-building workshop
    - We are organizing a capacity-building workshop in the beginning of June, where we will teach the basics of energy system modelling
    - As part of that workshop, we provide a temporarily-limited GAMS license to all participants
    - You could use that license to run your models
    - There's no event page yet (the organization needs to be finalized), but if you share your email address with us, we can send you an invitation in about one to two weeks
3. Apply for a GAMS test license yourself
    - By visiting https://www.gams.com/, you can apply for a free trial, which will grant you temporary access to a license which covers the CPLEX solver
    - They are usually quick to respond, so you might not have to wait more than 24 hours
    - I don't know how often you can apply for test licenses. If you organize some sort of course, the number might be higher, but you might also find that you face the same problem in three months or so without this third option.

**Reply by meng25meng** [reply]
Thank you for your answer. 1. I have applied for a free gurobi license before, can I replace this one?
I am currently a postgraduate student majoring in hydraulic engineering in Northwest A&F University. We are very interested in your conference. If I want to register for the conference, what are the requirements?

**Reply by glatterf42** [reply]
Gurobi is a solver, while GAMS is a modeling system which lets you use various solvers. Currently, we have hardcoded CPLEX as a solver, but you are very welcome to re-write that to select Gurobi instead, for which you can then use your license. However, I'd expect that you have to apply for a GAMS license, too, though you could mention in your request that you already have a license for Gurobi. 
There are no requirements per-se to join the capacity-building workshop :)
You should fulfill the [requirements for working with message_ix](https://docs.messageix.org/en/stable/prereqs.html), but that's about it. If you could please reach out to [me](https://iiasa.ac.at/staff/fridolin-glatter) and send me your name, email, and timezone, I can add you to the list of potential participants, so that you will receive the invitation whenever it's sent out.

**Reply by meng25meng** [reply]
Hello, I have received your reply. Based on your suggestion, I am modifying the multi-node model because I want to build a model with 31 nodes across the country. In order to avoid wasting the one-month trial period of the GAMS commercial license, I wanted to use the two-node model for testing first, but I found that the GAMS version did not match the API. Do you have any good solutions?
```python
[I 20:29:04.048 NotebookApp] Starting buffering for 91a82e6b-639b-4558-8e30-e787f1353e97:5c347ab33e3145899bd462953e68c611
[W 20:29:04.049 NotebookApp] zmq message arrived on closed channel
[I 20:29:05.063 NotebookApp] Restoring connection for 91a82e6b-639b-4558-8e30-e787f1353e97:5c347ab33e3145899bd462953e68c611
[I 20:29:05.129 NotebookApp] Replaying 1 buffered messages
[I 20:45:05.809 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 20:47:05.826 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 20:49:05.805 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 20:51:05.802 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:01:05.820 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:03:05.814 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:05:05.801 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:07:05.807 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:09:05.804 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:09:31.646 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:09:33.138 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:10:10.388 NotebookApp] Starting buffering for 91a82e6b-639b-4558-8e30-e787f1353e97:5c347ab33e3145899bd462953e68c611
[I 21:10:14.883 NotebookApp] Kernel shutdown: 91a82e6b-639b-4558-8e30-e787f1353e97
[I 21:10:33.374 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:10:35.361 NotebookApp] Starting buffering for d9e38d31-4338-4973-a9d9-ccdc2212699c:fcdcb938e5f54b4b8a531579bca5b08e
[I 21:10:42.464 NotebookApp] Kernel shutdown: d9e38d31-4338-4973-a9d9-ccdc2212699c
[I 21:10:51.895 NotebookApp] Kernel started: ec6a6a43-6f57-4d42-b11c-864d93d425bd, name: python3
0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
2024-04-11 21:12:18,364  INFO at.ac.iiasa.ixmp.Platform:165 - Welcome to the IX modeling platform!
2024-04-11 21:12:18,367  INFO at.ac.iiasa.ixmp.Platform:166 -  connected to database 'jdbc:hsqldb:file:C:/Users/Administrator/.local/share/ixmp/localdb/default' (user: ixmp)...
[I 21:12:51.885 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:18:51.875 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:20:51.888 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:22:52.670 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:24:51.880 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:26:51.883 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:28:51.886 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
[I 21:30:51.891 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
--- Warning: The GAMS version [45.7.0] differs from the API version [24.8.3].
[I 21:32:51.884 NotebookApp] Saving file at /iiasa-message_ix-f2bd088/tutorial/0122/BESMM-main/muilnode power system model.ipynb
```

**Reply by glatterf42** [reply]
This seems like a new topic, please always open new discussions/issues for new topics _after_ you've checked such discussions/issues don't already exist. 

This `Warning` is well known and happens with any GAMS version more recent than `24.8.3`, but this is not currently causing any failures in our test suite, so it should be fine for you as well.
